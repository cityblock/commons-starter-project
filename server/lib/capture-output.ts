import * as dotenv from 'dotenv';
dotenv.config();

/* tslint:disable no-var-requires */
const Logger = require('le_node');
/* tslint:enable no-var-requires */

export default class CaptureOutput {
  private releaseStdout: () => any;
  private releaseStderr: () => any;
  private logger: any;
  private tag: string | undefined;

  constructor(tag?: string) {
    this.logger = new Logger({
      token: process.env.LOGENTRIES_TOKEN,
    });
    this.tag = tag ? `[${tag}] - ` : '';
  }

  capture() {
    this.releaseStdout = this.captureStdout();
    this.releaseStderr = this.captureStderr();
  }

  release() {
    this.releaseStdout();
    this.releaseStderr();
  }

  private captureStream(stream: NodeJS.Socket, callback: (str: string) => any) {
    return this.hookStream(stream, (string: string) => {
      callback(`${this.tag}${string}`);
    });
  }

  private captureStdout() {
    return this.captureStream(process.stdout, this.logger.info.bind(this.logger));
  }

  private captureStderr() {
    return this.captureStream(process.stderr, this.logger.err.bind(this.logger));
  }

  // From https://gist.github.com/pguillory/729616#gistcomment-1946644
  private hookStream(stream: NodeJS.Socket, callback: any) {
    const old_write = stream.write;

    stream.write = (write =>
      (s: string): boolean => {
        write.apply(stream, [s]);
        callback(s);
        return true;
      })(stream.write);

    return () => (stream.write = old_write);
  }
}
