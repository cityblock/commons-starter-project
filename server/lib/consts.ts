export const adminTasksConcernTitle = 'Administrative Tasks';
export const attributionUserEmail = 'attribution.admin@cityblock.com';
export const attributionUserPermissions = 'black';
export const attributionUserUserRole = 'familyMember';
export const attributionUserClinicName = 'Attribution Clinic';
export const attributionUserClinicDepartmentId = 0;
