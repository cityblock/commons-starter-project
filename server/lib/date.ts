import { addDays, addHours, addMonths, addWeeks, addYears } from 'date-fns';
import { CompletedWithinInterval } from 'schema';

// helper for date math
export function dateAdd(
  date: Date,
  completedWithinNumber: number,
  completedWithinInterval: CompletedWithinInterval,
): Date {
  switch (completedWithinInterval) {
    case 'day':
      return addDays(date, completedWithinNumber);
    case 'hour':
      return addHours(date, completedWithinNumber);
    case 'month':
      return addMonths(date, completedWithinNumber);
    case 'year':
      return addYears(date, completedWithinNumber);
    case 'week':
      return addWeeks(date, completedWithinNumber);
    default:
      return date;
  }
}
