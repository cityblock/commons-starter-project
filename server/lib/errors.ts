export class AthenaResponseError extends Error {
  endpoint: string;
  status: number;

  constructor(message: string, endpoint: string, status: number) {
    super(message);
    this.endpoint = endpoint;
    this.status = status;
  }
}
