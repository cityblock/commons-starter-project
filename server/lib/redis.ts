import dotenv from 'dotenv';
import kue from 'kue';
import redis from 'redis';
import config from '../config';

dotenv.config();

interface IRedisConnection {
  createClientFactory: () => kue.redisClientFactory.RedisClient;
}

// Helper to generate correct redis client connection
export function createRedisClient(): IRedisConnection {
  return {
    createClientFactory: () =>
      redis.createClient(config.REDIS_PORT, config.REDIS_HOST, {
        password: config.REDIS_PASSWORD,
        tls:
          config.NODE_ENV === 'development'
            ? null
            : {
                port: config.REDIS_PORT,
                host: config.REDIS_HOST,
              },
      }),
  };
}
