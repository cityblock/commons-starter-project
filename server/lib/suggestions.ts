import { Transaction } from 'objection';
import CarePlanSuggestion from '../models/care-plan-suggestion';
import ConcernSuggestion from '../models/concern-suggestion';
import GoalSuggestion from '../models/goal-suggestion';

export const createSuggestions = async (patientId: string, txn: Transaction) => {
  const newConcernSuggestions = await ConcernSuggestion.getNewForPatient(patientId, txn);
  const newGoalSuggestions = await GoalSuggestion.getNewForPatient(patientId, txn);
  const formattedConcernSuggestions = newConcernSuggestions.map(concernSuggestion => ({
    patientId,
    suggestionType: 'concern' as any,
    concernId: concernSuggestion.id,
  }));
  const formattedGoalSuggestions = newGoalSuggestions.map(goalSuggestion => ({
    patientId,
    suggestionType: 'goal' as any,
    goalSuggestionTemplateId: goalSuggestion.id,
  }));

  const suggestions = formattedConcernSuggestions.concat(formattedGoalSuggestions as any);

  if (suggestions.length) {
    await CarePlanSuggestion.createMultiple({ suggestions }, txn);
  }
  return suggestions;
};

export const createSuggestionsForPatientScreeningToolSubmission = async (
  patientId: string,
  patientScreeningToolSubmissionId: string,
  txn?: Transaction,
) => {
  /* tslint:disable:max-line-length */
  const newConcernSuggestions = await ConcernSuggestion.getNewSuggestionsForPatientScreeningToolSubmission(
    patientId,
    patientScreeningToolSubmissionId,
    txn,
  );
  const newGoalSuggestions = await GoalSuggestion.getNewSuggestionsForPatientScreeningToolSubmission(
    patientId,
    patientScreeningToolSubmissionId,
    txn,
  );
  /* tslint:enable:max-line-length */
  const formattedConcernSuggestions = newConcernSuggestions.map(concernSuggestion => ({
    patientId,
    suggestionType: 'concern' as any,
    concernId: concernSuggestion.id,
    patientScreeningToolSubmissionId,
  }));
  const formattedGoalSuggestions = newGoalSuggestions.map(goalSuggestion => ({
    patientId,
    suggestionType: 'goal' as any,
    goalSuggestionTemplateId: goalSuggestion.id,
    patientScreeningToolSubmissionId,
  }));

  const suggestions = formattedConcernSuggestions.concat(formattedGoalSuggestions as any);
  if (suggestions.length) {
    await CarePlanSuggestion.createMultiple({ suggestions }, txn);
  }
  return suggestions;
};
