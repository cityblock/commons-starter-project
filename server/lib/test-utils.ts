import Knex from 'knex';
import omit from 'lodash/omit';
import Objection from 'objection';
import * as knexConfig from '../models/knexfile';

export const setupDb = () => {
  const config = (knexConfig as { [key: string]: any }).test;
  const knex = Knex(config);
  Objection.Model.knex(knex);
  return {
    destroy: () => {
      knex.destroy();
    },
  };
};

export const testGraphqlContext = (ctx: any) => ({
  getOrCreateTransaction: (
    testTransaction: Objection.Transaction,
    cb: <T>(cb: Objection.Transaction) => Promise<T>,
  ) => cb(testTransaction),
  ...ctx,
});

export const filterTimestamps = (object: { createdAt?: any, updatedAt?: any, deletedAt: any }) => 
  omit(object, 'createdAt', 'updatedAt', 'deletedAt');
