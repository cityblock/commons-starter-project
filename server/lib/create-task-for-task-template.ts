import { Transaction } from 'objection';
import { dateAdd } from '../lib/date';
import CareTeam from '../models/care-team';
import Task from '../models/task';
import TaskEvent from '../models/task-event';
import TaskTemplate from '../models/task-template';

export async function createTaskForTaskTemplate(
  taskTemplate: TaskTemplate,
  userId: string,
  patientId: string,
  txn: Transaction,
  patientGoalId?: string,
) {
  const { careTeamAssigneeRole, completedWithinInterval, completedWithinNumber } = taskTemplate;

  let dueAt: string | undefined;
  let assignedToId: string | undefined;

  const careTeam = await CareTeam.getForPatient(patientId);

  if (careTeamAssigneeRole) {
    const assignedCareTeamMember = careTeam.find(
      careTeamMember => careTeamMember.userRole === careTeamAssigneeRole,
    );

    if (assignedCareTeamMember) {
      assignedToId = assignedCareTeamMember.id;
    }
  }

  if (completedWithinInterval && completedWithinNumber) {
    dueAt = dateAdd(
      new Date(Date.now()),
      completedWithinNumber,
      completedWithinInterval,
    ).toISOString();
  }

  const task = await Task.create(
    {
      createdById: userId,
      title: taskTemplate.title,
      dueAt,
      patientId,
      assignedToId,
      patientGoalId,
    },
    txn,
  );

  await TaskEvent.create(
    {
      taskId: task.id,
      userId,
      eventType: 'create_task',
    },
    txn,
  );

  if (assignedToId) {
    await TaskEvent.create(
      {
        taskId: task.id,
        userId,
        eventType: 'edit_assignee',
        eventUserId: assignedToId,
      },
    txn,
    );
  }
}
