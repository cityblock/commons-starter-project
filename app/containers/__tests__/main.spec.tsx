import * as React from 'react';
import { MockedProvider } from 'react-apollo/lib/test-utils';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import MainContainer from '../main';

it('renders main container correctly', () => {
  const mockStore = configureMockStore([]);
  const tree = create(
    <MockedProvider mocks={[]} store={mockStore()}>
      <MainContainer />
    </MockedProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
