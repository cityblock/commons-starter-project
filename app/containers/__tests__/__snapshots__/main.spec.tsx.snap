// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders main container correctly 1`] = `
<div>
  <div
    className={undefined}
  />
  <div
    className={undefined}
  >
    A Sidewalk Labs urban health initiative
  </div>
</div>
`;
